### PR TITLE
Add ImageClassificationDataset for clearer requirements

### PR DESCRIPTION
**Pull request by zachgk**
This dataset adds some help to creating image classification datasets,
especially for users less familiar with DJL in DJL-Easy.

### CODE REVIEW

File: basicdataset/src/main/java/ai/djl/basicdataset/ImageClassificationDataset.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import ai.djl.training.dataset.Record;
+import java.io.IOException;
+
+/**
+ * A helper to create {@link ai.djl.training.dataset.Dataset}s for {@link
+ * ai.djl.Application.CV#IMAGE_CLASSIFICATION}.
+ */
+public abstract class ImageClassificationDataset extends RandomAccessDataset {
+
+    Image.Flag flag;
```
Comment (frankfliu) [inline basicdataset/src/main/java/ai/djl/basicdataset/ImageClassificationDataset.java:29]: Should this be private?

File: basicdataset/src/main/java/ai/djl/basicdataset/ImageClassificationDataset.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    public abstract Image getImage(ImageFactory imageFactory, long index) throws IOException;
+
+    /**
+     * Returns the class of the data item at the given index.
+     *
+     * @param index the index (if the dataset is a list of data items)
+     * @return the class number or the index into the list of classes of the desired class name
+     * @throws IOException if the data could not be loaded
+     */
+    public abstract long getClassNumber(long index) throws IOException;
```
Comment (frankfliu) [inline basicdataset/src/main/java/ai/djl/basicdataset/ImageClassificationDataset.java:59]: Should this return int? more that 2G classes?
```suggestion
    protected abstract int getClassNumber(long index) throws IOException;
```

File: basicdataset/src/main/java/ai/djl/basicdataset/ImageClassificationDataset.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * Returns the image at the given index in the dataset.
+     *
+     * @param imageFactory the imageFactory that can be used to create the {@link Image}
+     * @param index the index (if the dataset is a list of data items)
+     * @return the image
+     * @throws IOException if the image could not be loaded
+     */
+    public abstract Image getImage(ImageFactory imageFactory, long index) throws IOException;
```
Comment (frankfliu) [inline basicdataset/src/main/java/ai/djl/basicdataset/ImageClassificationDataset.java:50]: I think ImageFactory can be kept as member variable.
```suggestion
    protected abstract Image getImage(long index) throws IOException;
```

File: basicdataset/src/main/java/ai/djl/basicdataset/AbstractImageFolder.java [hunk 4]
```diff
@@ -64,14 +57,17 @@ protected AbstractImageFolder(ImageFolderBuilder<?> builder) {
...
-        NDArray array = ImageFactory.getInstance().fromFile(imagePath).toNDArray(manager, flag);
-        NDList d = new NDList(array);
-        NDList l = new NDList(manager.create(item.getValue()));
-        return new Record(d, l);
+        return imageFactory.fromFile(imagePath);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long getClassNumber(long index) throws IOException {
```
Comment (frankfliu) [inline basicdataset/src/main/java/ai/djl/basicdataset/AbstractImageFolder.java:68]:
```suggestion
    protected long getClassNumber(long index) {
```

File: basicdataset/src/main/java/ai/djl/basicdataset/ImageClassificationDataset.java (hunk 2 above)
Comment (zachgk) [reply basicdataset/src/main/java/ai/djl/basicdataset/ImageClassificationDataset.java:59]: I think long is a bit more standard as it corresponds to the size of an NDArray dimension. There may be cases involving people or content with this many classes as well

**Review by frankfliu — APPROVED** [review]

**Pull request merged**